autopush/webpush_server.py

```diff
@@ -0,0 +1,621 @@
+"""WebPush Server
+
+"""
+from threading import Thread
+from uuid import UUID, uuid4
+
+import attr
+from attr import (
+    attrs,
+    attrib,
+)
+from boto.dynamodb2.exceptions import ItemNotFound
+from boto.exception import JSONResponseError
+from typing import Dict, List, Optional  # noqa
+from twisted.logger import Logger
+
+from autopush.db import (  # noqa
+    DatabaseManager,
+    has_connected_this_month,
+    hasher,
+    generate_last_connect,
+)
+from autopush.config import AutopushConfig  # noqa
+from autopush.metrics import IMetrics  # noqa
+from autopush.types import JSONDict  # noqa
+from autopush.utils import WebPushNotification
+from autopush.websocket import USER_RECORD_VERSION
+from autopush_rs import AutopushCall, AutopushServer, AutopushQueue  # noqa
+
+log = Logger()
+
+# sentinel objects
+_STOP = object()
+
+
+# Conversion functions
+def uaid_from_str(input):
+    # type: (Optional[str]) -> Optional[UUID]
+    """Parse a uaid and verify the raw input matches the hex version (no
+    dashes)"""
+    try:
+        uuid = UUID(input)
+        if uuid.hex != input:
+            return None
+        return uuid
+    except (TypeError, ValueError):
+        return None
+
+
+@attrs(slots=True)
+class WebPushMessage(object):
+    """Serializable version of attributes needed for message delivery"""
+    uaid = attrib()  # type: str
+    timestamp = attrib()  # type: int
+    channelID = attrib()  # type: str
+    ttl = attrib()  # type: int
+    topic = attrib()  # type: str
+    version = attrib()  # type: str
+    sortkey_timestamp = attrib(default=None)  # type: Optional[int]
+    data = attrib(default=None)  # type: Optional[str]
+    headers = attrib(default=None)  # type: Optional[JSONDict]
+
+    @classmethod
+    def from_WebPushNotification(cls, notif):
+        # type: (WebPushNotification) -> WebPushMessage
+        p = notif.websocket_format()
+        del p["messageType"]
+        return cls(
+            uaid=notif.uaid.hex,
+            timestamp=int(notif.timestamp),
+            sortkey_timestamp=notif.sortkey_timestamp,
+            ttl=int(notif.ttl) if notif.ttl else notif.ttl,
```

Comment:
huh? not quite following this.

Reply:
else it will be whatever is there that evaluates to False, which is not actually good now that I read it cause a ttl of 0 evaluates to false. doh.